Bluetooth: audio: implement bap broadcast assistant long read/write

kruithofa opened the pull request:
Implements the long read/long write for BAP broadcast assistant.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/63362
fixes https://github.com/zephyrproject-rtos/zephyr/issues/64183

Commit:
Bluetooth: audio: bap: long write

Increase the buffer size to allow for long writes

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: audio: test: add test for long write

Added a test for long write, long read, and notifications

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: audio: bap: add support for long read

Make sure that buffers are large enough and add
semaphores so that we can do long reads

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: audio: bap: add support for long notifications

Add support for sending long notifications, coming
from a long read, back to the application

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,15 +192,148 @@ static int parse_recv_state(const void *data, uint16_t length,
+
+	LOG_DBG("conn %p err 0x%02x len %u", conn, err, length);
+
+	if (err) {
+		LOG_DBG("Failed to read: %u", err);
+
+		return BT_GATT_ITER_STOP;
+	}
```
Missing reset of the buffer

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,15 +192,148 @@ static int parse_recv_state(const void *data, uint16_t length,
+
+		broadcast_assistant.busy = false;
+		/* store data*/
+		net_buf_simple_add_mem(buf, data, length);
+		return parse_and_send_recv_state(conn, buf, buf->len, &recv_state);
```
If `parse_and_send_recv_state` fails, then we are also missing a reset of the buffer. 

Effectively the buffer needs to be reset on read complete or read error (whenever the read procedure ends). 

Ideally we should not call `parse_and_send_recv_state` until after we've read everything; else we are potentially sending partial receive states to the application.

kruithofa replied:
So bt_gatt_read can call us before it has received all the data? How  can we know that all the data has been received?

Thalley replied:
> So bt_gatt_read can call us before it has received all the data? How can we know that all the data has been received?

Yes, for long reads, the read callback is called for each segment. You know that all data has been read when `err == 0` and `data == NULL`

Commit:
Bluetooth: audio: fix bug in scan delegator sync

In the scan delegator when modifying the source we only want to
send a sync request to the upper layers if we are not yet synced,
and not also on a state change

Note that without this change the babblesim test for long writes
will fail

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Scan del bugfix

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on tests/bsim/bluetooth/audio/src/bap_broadcast_assistant_test.c:
```diff
@@ -424,6 +430,38 @@ static void test_bass_mod_source(void)
+	printk("Server PA synced\n");
+}
+
+void test_bass_long_write(void)
```
```suggestion
void test_bass_mod_source_long_meta(void)
```

Thalley replied:
We don't actually need to do `mod_src` twice - We could just have added the long meta to the existing `test_bass_mod_source`

kruithofa replied:
I considered that, and I suppose it would have been good enough since we test the normal read already in the other tests. Leaving it this way so it is obvious that we explicitly test the long read/writes

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -57,14 +59,19 @@ struct bap_broadcast_assistant_instance {
 static struct bt_uuid_16 uuid = BT_UUID_INIT_16(0);
 
-NET_BUF_SIMPLE_DEFINE_STATIC(cp_buf, CONFIG_BT_L2CAP_TX_MTU);
+#define CP_BUF_SIZE BT_ATT_MAX_ATTRIBUTE_LEN
+NET_BUF_SIMPLE_DEFINE_STATIC(cp_buf, CP_BUF_SIZE);
```
```suggestion
NET_BUF_SIMPLE_DEFINE_STATIC(att_buf, BT_ATT_MAX_ATTRIBUTE_LEN);
```

Suggest to rename it as it is no longer just used for the control points (CP) but all GATT/ATT operations

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+	int err;
+	int16_t index;
+
+	err = parse_recv_state(data, length, recv_state);
+
+	if (err != 0) {
```
```suggestion
	err = parse_recv_state(data, length, recv_state);
	if (err != 0) {
```

We genrally do not have empty lines between a value assignment and a verification check

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+		return BT_GATT_ITER_STOP;
+	}
+
+	index = lookup_index_by_handle(broadcast_assistant.long_read_handle);
```
Consider adding the handle to the function parameter instead of assigning the `long_read_handle` with the notification handle (purely for semantic purposes)

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
 	return 0;
 }
 
+static void bap_ntf_read_done(void)
```
Since this is (or could be) used for long reads without notification and the write operation, you could call it `bap_long_op_reset` or `bap_long_op_done`

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+		LOG_DBG("Failed to read: %u", err);
+		memset(read, 0, sizeof(*read));
+		bap_ntf_read_done();
+		bt_conn_unref(conn);
```
Why are you `unref`ing the `conn` here?

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+				length + cp_buf.len);
+			memset(read, 0, sizeof(*read));
+			bap_ntf_read_done();
+			bt_conn_unref(conn);
```
Ditto with the `unref`

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+	/* do the parse and callback to send  notify to application*/
+	parse_and_send_recv_state(conn, cp_buf.data, data_length, &recv_state);
+
+	bt_conn_unref(conn);
```
Ditto with the `unref`

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+	bt_conn_unref(conn);
+
+	return BT_GATT_ITER_STOP;
+}
+
+
+static void long_bap_read(struct bt_conn *conn, uint16_t handle)
```
```suggestion
}

static void long_bap_read(struct bt_conn *conn, uint16_t handle)
```

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+		if (err < 0) {
+			LOG_DBG("Failed to reschedule read work: %d", err);
+			bap_ntf_read_done();
+			bt_conn_unref(conn);
```
Ditto with the `unref`

I think it makes sense to do 1 `bt_conn_ref` when you initate the discover and store the `conn` pointer in the instance, and then 1 `bt_conn_unref` when you disconnect or the discover function is called again. 

See e.g. the similar changes here: https://github.com/zephyrproject-rtos/zephyr/pull/67629/files#diff-a1827bc1ecbe96b568cf0cc5ea28bf55f8593f2ae005a9c65f6f051254b21954

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -333,6 +484,16 @@ static uint8_t read_recv_state_cb(struct bt_conn *conn, uint8_t err,
 }
 
+static void discover_init(void)
+{
+
+	(void)memset(&broadcast_assistant, 0, sizeof(broadcast_assistant));
```
```suggestion
{
	(void)memset(&broadcast_assistant, 0, sizeof(broadcast_assistant));
```

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -475,6 +636,11 @@ static void bap_broadcast_assistant_write_cp_cb(struct bt_conn *conn, uint8_t er
 		return;
 	}
 
+	/*
+	 * we reset the buffer, so that we are ready for new notifications and writes
+	 */
```
```suggestion
	/* we reset the buffer, so that we are ready for new notifications and writes */
```

Thalley commented on subsys/bluetooth/audio/bap_scan_delegator.c:
```diff
@@ -137,11 +140,29 @@ static void bt_debug_dump_recv_state(const struct bass_recv_state_internal *recv
+
+	LOG_DBG("Sending bytes %d", ntf_size);
+	err = bt_gatt_notify_uuid(NULL, BT_UUID_BASS_RECV_STATE,
 				      internal_state->attr, read_buf.data,
-				      read_buf.len);
+				      ntf_size);
```
Indentation is off here

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -184,14 +191,152 @@ static int parse_recv_state(const void *data, uint16_t length,
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	/*
+	 * we reset the buffer so that it is ready for new data
+	 */
```
```suggestion
	/* reset the buffer so that it is ready for new data */
```

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -206,30 +351,28 @@ static uint8_t notify_handler(struct bt_conn *conn,
-		    broadcast_assistant_cbs->recv_state != NULL) {
-			broadcast_assistant_cbs->recv_state(conn, 0, &
-							    recv_state);
+		broadcast_assistant.long_read_handle = handle;
+		if (length == max_ntf_size) {
+			if (!broadcast_assistant.busy) {
+				net_buf_simple_add_mem(&att_buf, data, length);
+			}
+			long_bap_read(conn, handle);
```
Hmm, if we are in the middle of a long read or write (i.e. `broadcast_assistant.busy` is true), then we shouldn't overwrite `broadcast_assistant.long_read_handle = handle;`. 

I guess we may actually need to store a list or array of pending handles to be read.

kruithofa replied:
Hmm, do we have the same problem in the bap_unicast_client?

Thalley replied:
No - It has been solved there as we have a timer for each ASE. We may need the same approach here with a timer per receive state

Thalley reviewed (changes requested):
Looks pretty good now - But I think we may still have a corner case that isn't handled.

fredrikdanebjer commented on subsys/bluetooth/audio/Kconfig.bap:
```diff
@@ -235,6 +235,15 @@ config BT_BAP_SCAN_DELEGATOR_MAX_SUBGROUPS
+	default 50
+	help
+	  The number of milliseconds that the scan delegator implementation will maximum wait
+	  before rejecting a ASE read or dropping a notification if the scan delegator state is
```
>before rejecting a ASE read or dropping a notification if the scan delegator state is

an ASE
